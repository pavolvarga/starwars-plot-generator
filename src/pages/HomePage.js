import 'bootstrap/dist/css/bootstrap.min.css';
import React, { Component } from 'react';
import { Col, FormGroup, Label, Input, Form, Button, Container } from 'reactstrap';
import { withRouter } from 'react-router-dom';

class InputForm extends Component {

    constructor(props) {

        super(props);

        this.state = {
           selected: {
               person: undefined,
               planet: undefined
           }
        };

        this.handleOnPersonInputChange = this.handleOnPersonInputChange.bind(this);
        this.handleOnPlanetInputChange = this.handleOnPlanetInputChange.bind(this);
    }

    handleOnPersonInputChange(person) {
        this.setState((state) => ({
            selected: {...state.selected, person}
        }));
    }

    handleOnPlanetInputChange(planet) {
        this.setState((state) => ({
            selected: {...state.selected, planet}
        }));
    }

    isGenerateBntDisabled() {
        return !(!!this.state.selected.person && !!this.state.selected.planet);
    }

    render() {
        return (
            <div>
                <div className="text-center">
                    <h1>Star Wars Plot Generator</h1>
                </div>
                <Container>
                    <Form>
                        <Col>
                            <FormGroup size="lg" row>
                                <Label for="input-person" size="lg">Character</Label>
                                <Input
                                    id="input-person"
                                    name="input-person"
                                    placeholder="Please enter a character"
                                    size="lg"
                                    onChange={e => this.handleOnPersonInputChange(e.target.value)} />
                            </FormGroup>
                        </Col>
                        <Col>
                            <FormGroup size="lg" row>
                                <Label for="input-planet" size="lg">Planet</Label>
                                <Input
                                    id="input-planet"
                                    name="input-planet"
                                    placeholder="Please enter a planet"
                                    size="lg"
                                    onChange={e => this.handleOnPlanetInputChange(e.target.value)} />
                            </FormGroup>
                        </Col>
                        <Col>
                            <div className="text-center">
                                <Button
                                    color="primary"
                                    size="lg"
                                    onClick={e => this.props.history.push('/plot')}
                                    disabled={this.isGenerateBntDisabled()}>
                                    Generate Plot
                                </Button>
                            </div>
                        </Col>
                    </Form>
                </Container>
            </div>
        )
    }
}

export { InputForm };
