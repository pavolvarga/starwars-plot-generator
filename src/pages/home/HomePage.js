import 'bootstrap/dist/css/bootstrap.min.css';
import React, { Component } from 'react';
import { Form, Container } from 'reactstrap';
import { loadStarWarsData } from './../../common/util';
import { StarWarsInput } from "./StarWarsInput";
import { GenerateBnt } from "./GenerateBnt";
import { OptionalInputs } from "./OptionalInputs";
import { OPTIONAL_RESOURCES, FAILED_LOAD_COOL_DOWN } from './../../common/const';
import { LoadFailedAlerts } from './LoadFailedAlerts';

const OPTIONAL_RESOURCES_SINGULAR = OPTIONAL_RESOURCES.map(r => r.singular);

function getPluralName(singularName) {
    return OPTIONAL_RESOURCES.filter(r => r.singular === singularName)[0].plural;
}

class InputForm extends Component {

    constructor(props) {

        super(props);

        this.state = {
            person: {visible: true, selected: undefined, data: props.customProps.people, loadingInProgress: false, loadFailed: false},
            planet: {visible: true, selected: undefined, data: props.customProps.planets, loadingInProgress: false, loadFailed: false},
            starship: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false},
            vehicle: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false},
            species: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false}
        };

        this.setSelectedValue = this.setSelectedValue.bind(this);
        this.toggleVisibility = this.toggleVisibility.bind(this);
        this.isGenerateBntDisabled = this.isGenerateBntDisabled.bind(this);
        this.generatePlot = this.generatePlot.bind(this);
        this.loadResourceData = this.loadResourceData.bind(this);
    }

    isGenerateBntDisabled() {
        return !(!!this.state.person.selected && !!this.state.planet.selected);
    }

    generatePlot() {
        const state = {};
        Object.keys(this.state).forEach(key => {
            state[key] = {visible: this.state[key].visible, selected: this.state[key].selected};
        });
        this.props.history.push({
            pathname: '/plot',
            state
        });
    }

    setSelectedValue(name, value) {
        const
            resource = this.state[name],
            updated = Object.assign(resource, {selected: value}),
            obj = {};

        obj[name] = updated;
        this.setState(state => Object.assign(state, obj));
    }

    loadResourceData(name) {
        const
            resourcePlural = getPluralName(name),
            resolve = (function storeOptionalResourceData(value) {

                const
                    resource = this.state[name],
                    obj = {},
                    updated = Object.assign(resource, {loadingInProgress: false, data: value});

                obj[name] = updated;
                this.setState(state => Object.assign(state, obj));

            }).bind(this),
            reject = (function updateLoadFailed (err) {
                const
                    resource = this.state[name],
                    obj = {},
                    updated = Object.assign(resource, {loadFailed: true, loadingInProgress: false});

                obj[name] = updated;
                this.setState(state => Object.assign(state, obj));

                //clear alert after specified time, and allow user to try it again
                setTimeout(function clearLoadFailed() {
                    const cleared = Object.assign(resource, {visible: false, loadFailed: false});
                    obj[name] = cleared;
                    this.setState(state => Object.assign(state, obj));
                }.bind(this), FAILED_LOAD_COOL_DOWN * 1000);

            }).bind(this);

        loadStarWarsData(resourcePlural, resolve, reject);
    }

    toggleVisibility(name) {

        const
            resource = this.state[name],
            obj = {};

        let updated = Object.assign(resource, {visible: !resource.visible});

        //load data for the input in other function, after the input became visible
        //if it wasn't already loaded
        if(!this.state[name].loadingInProgress && this.state[name].data.length === 0) {
            updated = Object.assign(updated, {loadingInProgress: true});
            setTimeout(() => this.loadResourceData(name), 1);
        }

        obj[name] = updated;
        this.setState(state => Object.assign(state, obj));
    }

    render() {

        const personProps = {
            id: "input-person",
            name: "input-person",
            label: "Character",
            placeholder: "Please enter a character",
            data: this.state.person.data,
            setFn: (function setPerson(value) {
                this.setSelectedValue('person', value);
            }).bind(this),
            visible: this.state.person.visible
        };
        const planetProps = {
            id: "input-planet",
            name: "input-planet",
            label: "Planet",
            placeholder: "Please enter a planet",
            data: this.state.planet.data,
            setFn: (function setPlanet(value) {
                this.setSelectedValue('planet', value);
            }).bind(this),
            visible: this.state.planet.visible
        };
        const starshipProps = {
            id: "input-starship",
            name: "input-starship",
            label: "Starship",
            placeholder: "Please enter a starship",
            data: this.state.starship.data,
            setFn: (function setStarship(value) {
                this.setSelectedValue('starship', value);
            }).bind(this),
            visible: this.state.starship.visible
        };
        const vehicleProps = {
            id: "input-vehicle",
            name: "input-vehicle",
            label: "Vehicle",
            placeholder: "Please enter a vehicle",
            data: this.state.vehicle.data,
            setFn: (function setVehicle(value) {
                this.setSelectedValue('vehicle', value);
            }).bind(this),
            visible: this.state.vehicle.visible
        };
        const speciesProps = {
            id: "input-species",
            name: "input-species",
            label: "Species",
            placeholder: "Please enter a species",
            data: this.state.species.data,
            setFn: (function setSpecies(value) {
                this.setSelectedValue('species', value);
            }).bind(this),
            visible: this.state.species.visible
        };

        const generateBntProps = {
            isGenerateBntDisabled: this.isGenerateBntDisabled,
            generatePlot: this.generatePlot
        };

        const optionalInputsProps = {
            toggleVisibilityFn: this.toggleVisibility,
            //an array with boolean values representing weather an input for an optional resource is visible or not
            visibles: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].visible),
            //an array with boolean values representing weather the load button is disabled
            disabled: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].loadFailed)
        };

        const alertsProps = {
            visibles: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].loadFailed)
        };

        return (
            <div>
                <Container>
                    <Form>
                        <StarWarsInput {...personProps} />
                        <StarWarsInput {...planetProps} />
                        <StarWarsInput {...starshipProps} />
                        <StarWarsInput {...vehicleProps} />
                        <StarWarsInput {...speciesProps} />
                        <OptionalInputs {...optionalInputsProps} />
                        <LoadFailedAlerts {...alertsProps} />
                        <GenerateBnt {...generateBntProps} />
                    </Form>
                </Container>
            </div>
        )
    }
}

export { InputForm };
