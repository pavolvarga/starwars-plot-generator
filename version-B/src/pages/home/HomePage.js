import 'bootstrap/dist/css/bootstrap.min.css';
import React, { Component } from 'react';
import { Form, Container } from 'reactstrap';
import { loadStarWarsData } from '../../common/load-data';
import { StarWarsSearch } from "./StarWarsSearch";
import { GenerateBnt } from "./GenerateBnt";
import { OptionalInputs } from "./OptionalInputs";
import { OPTIONAL_RESOURCES, FAILED_LOAD_COOL_DOWN } from './../../common/const';
import { LoadFailedAlerts } from './LoadFailedAlerts';

const OPTIONAL_RESOURCES_SINGULAR = OPTIONAL_RESOURCES.map(r => r.singular);

function getPluralName(singularName) {
    return OPTIONAL_RESOURCES.filter(r => r.singular === singularName)[0].plural;
}

function upperCase(text) {
    return text.charAt(0).toUpperCase() + text.slice(1);
}

class InputForm extends Component {

    constructor(props) {

        super(props);

        this.state = {
            person: {visible: true, selected: undefined, data: props.customProps.people, loadingInProgress: false, loadFailed: false},
            planet: {visible: true, selected: undefined, data: props.customProps.planets, loadingInProgress: false, loadFailed: false},
            starship: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false},
            vehicle: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false},
            species: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false}
        };

        this.setSelectedValue = this.setSelectedValue.bind(this);
        this.toggleVisibility = this.toggleVisibility.bind(this);
        this.isGenerateBntDisabled = this.isGenerateBntDisabled.bind(this);
        this.generatePlot = this.generatePlot.bind(this);
        this.loadResourceData = this.loadResourceData.bind(this);
        this.createInputProps = this.createInputProps.bind(this);
    }

    isGenerateBntDisabled() {
        return !(!!this.state.person.selected && !!this.state.planet.selected);
    }

    generatePlot() {
        const state = {};
        Object.keys(this.state).forEach(key => {
            state[key] = {selected: this.state[key].selected};
        });
        this.props.history.push({
            pathname: '/plot',
            state
        });
    }

    setSelectedValue(name, value) {
        const
            resource = this.state[name],
            updated = Object.assign(resource, {selected: value}),
            obj = {};

        obj[name] = updated;
        this.setState(state => Object.assign(state, obj));
    }

    loadResourceData(name) {
        const
            resourcePlural = getPluralName(name),
            resolve = (function storeOptionalResourceData(value) {

                const
                    resource = this.state[name],
                    obj = {},
                    updated = Object.assign(resource, {loadingInProgress: false, data: value});

                obj[name] = updated;
                this.setState(state => Object.assign(state, obj));

            }).bind(this),
            reject = (function updateLoadFailed (err) {
                const
                    resource = this.state[name],
                    obj = {},
                    updated = Object.assign(resource, {loadFailed: true, loadingInProgress: false});

                obj[name] = updated;
                this.setState(state => Object.assign(state, obj));

                //clear alert after specified time, and allow user to try it again
                setTimeout(function clearLoadFailed() {
                    const cleared = Object.assign(resource, {visible: false, loadFailed: false});
                    obj[name] = cleared;
                    this.setState(state => Object.assign(state, obj));
                }.bind(this), FAILED_LOAD_COOL_DOWN * 1000);

            }).bind(this);

        loadStarWarsData(resourcePlural, resolve, reject);
    }

    toggleVisibility(name) {

        const
            resource = this.state[name],
            obj = {};

        let updated = Object.assign(resource, {visible: !resource.visible, selected: undefined});

        //load data for the input in other function, after the input became visible
        //if it wasn't already loaded
        if(!this.state[name].loadingInProgress && this.state[name].data.length === 0) {
            updated = Object.assign(updated, {loadingInProgress: true});
            setTimeout(() => this.loadResourceData(name), 1);
        }

        obj[name] = updated;
        this.setState(state => Object.assign(state, obj));
    }

    createInputProps(name, label) {

        const
            normalizedLabel = label ? label : name,
            upperCasedLabel = upperCase(normalizedLabel);

        return {
            id: `input-${name}`,
            name: `input-${name}`,
            label: upperCasedLabel,
            placeholder: `Please enter a ${normalizedLabel}`,
            data: this.state[name].data,
            setFn: (function setEnteredValue(value) {
                this.setSelectedValue(name, value);
            }).bind(this),
            visible: this.state[name].visible
        };
    }

    render() {

        const
            personProps = this.createInputProps('person', 'character'),
            planetProps = this.createInputProps('planet'),
            starshipProps = this.createInputProps('starship'),
            vehicleProps = this.createInputProps('vehicle'),
            speciesProps = this.createInputProps('species');

        const generateBntProps = {
            isGenerateBntDisabled: this.isGenerateBntDisabled,
            generatePlot: this.generatePlot
        };

        const optionalInputsProps = {
            toggleVisibilityFn: this.toggleVisibility,
            //an array with boolean values representing weather an input for an optional resource is visible or not
            visibles: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].visible),
            //an array with boolean values representing weather the load button is disabled
            disabled: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].loadFailed)
        };

        const alertsProps = {
            visibles: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].loadFailed)
        };

        return (
            <div>
                <Container>
                    <Form>
                        <StarWarsSearch {...personProps} />
                        <StarWarsSearch {...planetProps} />
                        <StarWarsSearch {...starshipProps} />
                        <StarWarsSearch {...vehicleProps} />
                        <StarWarsSearch {...speciesProps} />
                        <OptionalInputs {...optionalInputsProps} />
                        <LoadFailedAlerts {...alertsProps} />
                        <GenerateBnt {...generateBntProps} />
                    </Form>
                </Container>
            </div>
        )
    }
}

export { InputForm };
