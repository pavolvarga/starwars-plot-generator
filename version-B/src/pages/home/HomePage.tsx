import 'bootstrap/dist/css/bootstrap.min.css';
import React, { Component } from 'react';
import { Form, Container } from 'reactstrap';
import { loadStarWarsData, ResourceData } from '../../common/load-data';
import { StarWarsSearch, Suggestion, StarWarsSearchProps } from "./StarWarsSearch";
import { GenerateBnt } from "./GenerateBnt";
import { OptionalInputs } from "./OptionalInputs";
import { OPTIONAL_RESOURCES, FAILED_LOAD_COOL_DOWN } from '../../common/const';
import { LoadFailedAlerts } from './LoadFailedAlerts';

const OPTIONAL_RESOURCES_SINGULAR: string[] = OPTIONAL_RESOURCES.map(r => r.singular);

function getPluralName(singularName: string): string {
    return OPTIONAL_RESOURCES.filter(r => r.singular === singularName)[0].plural;
}

function upperCase(text: string): string {
    return text.charAt(0).toUpperCase() + text.slice(1);
}

type InputState = {
    visible: boolean,
    selected: Suggestion | undefined,
    data: ResourceData[],
    loadingInProgress?: boolean,
    loadFailed: boolean
};
type InputFormState = {
    person: InputState,
    planet: InputState,
    starship: InputState,
    vehicle: InputState,
    species: InputState,
};
class InputForm extends Component<any, InputFormState> {

    constructor(props) {

        super(props);

        this.state = {
            person: {visible: true, selected: undefined, data: props.customProps.people, loadingInProgress: false, loadFailed: false},
            planet: {visible: true, selected: undefined, data: props.customProps.planets, loadingInProgress: false, loadFailed: false},
            starship: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false},
            vehicle: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false},
            species: {visible: false, selected: undefined, data: [], loadingInProgress: false, loadFailed: false}
        };

        this.setSelectedValue = this.setSelectedValue.bind(this);
        this.toggleVisibility = this.toggleVisibility.bind(this);
        this.isGenerateBntDisabled = this.isGenerateBntDisabled.bind(this);
        this.generatePlot = this.generatePlot.bind(this);
        this.loadResourceData = this.loadResourceData.bind(this);
        this.createSearchInputProps = this.createSearchInputProps.bind(this);
    }

    isGenerateBntDisabled(): boolean {
        return !(!!this.state.person.selected && !!this.state.planet.selected);
    }

    generatePlot(): void {
        const state = {};
        Object.keys(this.state).forEach(key => {
            state[key] = {selected: this.state[key].selected};
        });
        this.props.history.push({
            pathname: '/plot',
            state
        });
    }

    /**
     * @param name - the name of input (person, planet, ...)
     * @param suggestion - either undefined (if not selected) or a selected suggestion value
     */
    setSelectedValue(name: string, suggestion: Suggestion | undefined): void {
        const
            resource = this.state[name],
            updated = Object.assign(resource, {selected: suggestion}),
            obj = {};

        obj[name] = updated;
        this.setState(state => Object.assign(state, obj));
    }

    loadResourceData(name): void {
        const
            resourcePlural = getPluralName(name),
            resolve = (function storeOptionalResourceData(this: InputForm, value) {

                const
                    resource = this.state[name],
                    obj = {},
                    updated = Object.assign(resource, {loadingInProgress: false, data: value});

                obj[name] = updated;
                this.setState(state => Object.assign(state, obj));

            }).bind(this),
            reject = (function updateLoadFailed (this: InputForm, err) {
                const
                    resource = this.state[name],
                    obj = {},
                    updated = Object.assign(resource, {loadFailed: true, loadingInProgress: false});

                obj[name] = updated;
                this.setState(state => Object.assign(state, obj));

                //clear alert after specified time, and allow user to try it again
                setTimeout(function clearLoadFailed(this: InputForm) {
                    const cleared = Object.assign(resource, {visible: false, loadFailed: false});
                    obj[name] = cleared;
                    this.setState(state => Object.assign(state, obj));
                }.bind(this), FAILED_LOAD_COOL_DOWN * 1000);

            }).bind(this);

        loadStarWarsData(resourcePlural, resolve, reject);
    }

    toggleVisibility(name: string): void {

        const
            resource = this.state[name],
            obj = {};

        let updated = Object.assign(resource, {visible: !resource.visible, selected: undefined});

        //load data for the input in other function, after the input became visible
        //if it wasn't already loaded
        if(!this.state[name].loadingInProgress && this.state[name].data.length === 0) {
            updated = Object.assign(updated, {loadingInProgress: true});
            setTimeout(() => this.loadResourceData(name), 1);
        }

        obj[name] = updated;
        this.setState(state => Object.assign(state, obj));
    }

    createSearchInputProps(name: string, label?: string): StarWarsSearchProps {

        const
            normalizedLabel = label ? label : name,
            upperCasedLabel = upperCase(normalizedLabel);

        return {
            id: `input-${name}`,
            name: `input-${name}`,
            label: upperCasedLabel,
            placeholder: `Please enter a ${normalizedLabel}`,
            data: this.state[name].data,
            setFn: (function setEnteredValue(this: InputForm, selectedSuggestion: Suggestion | undefined) {
                this.setSelectedValue(name, selectedSuggestion);
            }).bind(this),
            visible: this.state[name].visible
        };
    }

    render() {

        const
            personProps = this.createSearchInputProps('person', 'character'),
            planetProps = this.createSearchInputProps('planet'),
            starshipProps = this.createSearchInputProps('starship'),
            vehicleProps = this.createSearchInputProps('vehicle'),
            speciesProps = this.createSearchInputProps('species');

        const generateBntProps = {
            isGenerateBntDisabled: this.isGenerateBntDisabled,
            generatePlot: this.generatePlot
        };

        const optionalInputsProps = {
            toggleVisibilityFn: this.toggleVisibility,
            //an array with boolean values representing weather an input for an optional resource is visible or not
            visibles: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].visible),
            //an array with boolean values representing weather the load button is disabled
            disabled: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].loadFailed)
        };

        const alertsProps = {
            visibles: OPTIONAL_RESOURCES_SINGULAR.map(x => this.state[x].loadFailed)
        };

        return (
            <div>
                <Container>
                    <Form>
                        <StarWarsSearch {...personProps} />
                        <StarWarsSearch {...planetProps} />
                        <StarWarsSearch {...starshipProps} />
                        <StarWarsSearch {...vehicleProps} />
                        <StarWarsSearch {...speciesProps} />
                        <OptionalInputs {...optionalInputsProps} />
                        <LoadFailedAlerts {...alertsProps} />
                        <GenerateBnt {...generateBntProps} />
                    </Form>
                </Container>
            </div>
        )
    }
}

export { InputForm };
