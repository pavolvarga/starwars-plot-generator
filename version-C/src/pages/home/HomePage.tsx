import React, { FC, useContext, useEffect, Fragment } from 'react';

import { AppContext } from "../../AppContext";
import { AppState } from "../../common/types";
import { InputForm } from "./InputForm";
import { Loader } from './Loader';

const HomePage: FC<any> = (props: any) => {

    const
        context = useContext(AppContext),
        {loadMandatoryResourceData} = (context as AppState);

    useEffect(() => {
        loadMandatoryResourceData();
        // eslint-disable-next-line
    }, []);

    return (
        <Fragment>
            <Loader/>
            <InputForm/>
        </Fragment>
    );
};

export { HomePage };