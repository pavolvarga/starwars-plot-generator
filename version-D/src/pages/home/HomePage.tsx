import React, { FC, useContext, useEffect, Fragment } from 'react';

import { AppContext } from "../../AppContext";
import { AppState } from "../../common/types";
import { InputForm } from "./InputForm";
import { Loader } from './Loader';
import { RouteComponentProps } from "react-router";
import  {Alert } from "reactstrap";

const InitialLoadFailedAlert: FC = () => {
    return (
        <div className="text-center">
            <Alert color="danger">
                Loading of initial data failed, please try again later.
            </Alert>
        </div>
    );
};

const HomePage: FC<RouteComponentProps> = (props: RouteComponentProps) => {

    const
        context = useContext(AppContext),
        {loadMandatoryResourceData, hasLoadingOfMandatoryDataFailed} = (context as AppState);

    //load mandatory resources at component did mount time
    useEffect(() => {
        loadMandatoryResourceData();
        // eslint-disable-next-line
    }, []);

    if (hasLoadingOfMandatoryDataFailed()) {
        return <InitialLoadFailedAlert/>
    }

    return (
        <Fragment>
            <Loader />
            <InputForm {...props} />
        </Fragment>
    );
};

export { HomePage };