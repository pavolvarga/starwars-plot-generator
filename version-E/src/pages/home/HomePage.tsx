import React, { FC } from 'react';
import { RouteComponentProps } from "react-router";

import { Loader } from './Loader';
import { InputForm } from "./InputForm";
import { InitialLoadFailedAlert } from "./InitialLoadFailedAlert";

export const HomePage: FC<RouteComponentProps> = (props: RouteComponentProps) => {
    return (
        <div className="container">
            <InitialLoadFailedAlert />
            <Loader />
            <InputForm {...props} />
        </div>
    );
};
