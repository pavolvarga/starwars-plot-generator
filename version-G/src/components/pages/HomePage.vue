<template>
  <form v-if="isMandatoryDataLoaded()">
    <star-wars-search
      v-for="input in getVisibleInputs()"
      :key="input.name"
      :name="input.name"
      :label="input.label"
      :data="input.data">
    </star-wars-search>
    <optional-inputs v-if="getVisibleInputs().length > 0"></optional-inputs>
    <load-failed-alerts></load-failed-alerts>
    <the-generate-button></the-generate-button>
  </form>
  <div class="row" v-if="hasLoadingOfMandatoryDataFailed()">
    <div class="col-lg-12">
      <div class="alert alert-danger" role="alert">
        Loading of mandatory data failed, please try again later.
      </div>
    </div>
  </div>
  <div class="row" v-if="!isMandatoryDataLoaded() && !hasLoadingOfMandatoryDataFailed()">
    <div class="col-md-12">
      <h3>Please wait while star wars data is being loaded</h3>
      <div class="spinner-border" role="status">
        <span class="sr-only">Loading...</span>
      </div>
    </div>
  </div>
</template>

<script>
import { defineComponent } from 'vue';
import { useStore } from 'vuex'
import { key } from "@/store/store";
import StarWarsSearch from "@/components/StarWarsSearch";
import OptionalInputs from "@/components/OptionalInputs";
import LoadFailedAlerts from "@/components/LoadFailedAlerts";
import TheGenerateButton from "@/components/TheGenerateButton";

export default defineComponent({
  name: 'HomePage',
  components: {
    StarWarsSearch,
    OptionalInputs,
    LoadFailedAlerts,
    TheGenerateButton,
  },
  setup() {
    const store = useStore(key);
    // if returning from the /result page, reset selected values to start from the beginning
    store.dispatch("resetSelected");
    store.dispatch("loadMandatoryResources");

    function getVisibleInputs() {
      return store.getters.getVisibleInputs;
    }
    function isMandatoryDataLoaded() {
      return store.getters.isMandatoryDataLoaded;
    }
    function hasLoadingOfMandatoryDataFailed() {
      return store.getters.hasLoadingOfMandatoryDataFailed;
    }

    return {
      hasLoadingOfMandatoryDataFailed,
      isMandatoryDataLoaded,
      getVisibleInputs
    }
  }
})
</script>

<style scoped>
input {
  display: block;
  width: 100%;
}
.spinner-border {
  width: 5rem;
  height: 5rem;
}
</style>
