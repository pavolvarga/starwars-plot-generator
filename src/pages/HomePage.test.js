import React from 'react';
import { shallow } from 'enzyme';

import { InputForm } from './HomePage';

describe('HomePage', () => {
    it('should render correctly in "debug" mode', () => {

        const component = shallow(<InputForm debug />);

        expect(component).toMatchSnapshot();
    });

    describe('generate button dissabled propery', () => {
       it('should be false if input values are set', () => {

           const component = shallow(<InputForm/>);
           component.setState({
               selected: {
                   person: 'Luke Skywalker',
                   planet: 'Tatooine'
               }
           });

           const result = component.instance().isGenerateBntDisabled();

           expect(result).toEqual(false);
       });
       it('should be true if one input value is undefined', () => {

           const component = shallow(<InputForm/>);
           component.setState({
               selected: {
                   person: undefined,
                   planet: 'Tatooine'
               }
           });

           const result = component.instance().isGenerateBntDisabled();

           expect(result).toEqual(true);
       });
        it('should be true if one input value is an empty string', () => {

            const component = shallow(<InputForm/>);
            component.setState({
                selected: {
                    person: '',
                    planet: 'Tatooine'
                }
            });

            const result = component.instance().isGenerateBntDisabled();

            expect(result).toEqual(true);
        });
    });
});