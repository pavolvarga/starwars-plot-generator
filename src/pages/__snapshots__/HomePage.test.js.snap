// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render correctly in "debug" mode 1`] = `
<div>
  <div
    className="text-center"
  >
    <h1>
      Star Wars Plot Generator
    </h1>
  </div>
  <Container
    tag="div"
  >
    <Form
      tag="form"
    >
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <FormGroup
          row={true}
          size="lg"
          tag="div"
        >
          <Label
            for="input-person"
            size="lg"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Character
          </Label>
          <Input
            id="input-person"
            name="input-person"
            onChange={[Function]}
            placeholder="Please enter a character"
            size="lg"
            type="text"
          />
        </FormGroup>
      </Col>
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <FormGroup
          row={true}
          size="lg"
          tag="div"
        >
          <Label
            for="input-planet"
            size="lg"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            Planet
          </Label>
          <Input
            id="input-planet"
            name="input-planet"
            onChange={[Function]}
            placeholder="Please enter a planet"
            size="lg"
            type="text"
          />
        </FormGroup>
      </Col>
      <Col
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <div
          className="text-center"
        >
          <Button
            color="primary"
            disabled={true}
            onClick={[Function]}
            size="lg"
            tag="button"
          >
            Generate Plot
          </Button>
        </div>
      </Col>
    </Form>
  </Container>
</div>
`;
